### PR TITLE
Support new Intro Offers Endpoint

### PR DESCRIPTION
**Pull request by cbauerman**
#### Changes proposed in this Pull Request

- Support new `/introductory-offers` endpoint ( D72857-code )
- Use introductory offers to display discount prices on the variant picker in the cart

| Before ( introductory offer not displayed on variant ) | After ( introductory offer displayed on variant ) |
| ------------------------------------------------------ | --------------------------------------------------------|
|  |  |

#### Testing instructions

1. Add an assortment of Jetpack annual plans and products to the cart
2. Click "Edit" in cart to open the product variant selector
2. Verify that in the product variant picker the correct introductory offer ( 50% off ).
3. Verify the cart total is correct.
4. Add an annual product to the cart via `cloud.jetpack.com/pricing` in a logged-out/incognito window.
5. Verify the correct discount and price is shown on the variant selector.

Related to 1201295898168937-as-1201646705905156

### CODE REVIEW

File: client/state/action-types.ts [hunk 1]
```diff
@@ -865,6 +865,10 @@ export const SITE_FEATURES_FETCH_COMPLETED = 'SITE_FEATURES_FETCH_COMPLETED';
 export const SITE_FEATURES_FETCH_FAILED = 'SITE_FEATURES_FETCH_FAILED';
 export const SITE_FEATURES_REMOVE = 'SITE_FEATURES_REMOVE';
 export const SITE_FRONT_PAGE_UPDATE = 'SITE_FRONT_PAGE_UPDATE';
+export const SITE_INTRO_OFFER_RECEIVE = 'SITE_INTRO_OFFER_RECEIVE';
```
Comment (cbauerman) [inline client/state/action-types.ts:860]: I opted for a traditional query/redux pattern over `react-query` because the cart is heavily redux based so this fits better.

File: client/state/selectors/get-is-eligible-for-intro-offer.ts [hunk 2]
```diff
@@ -0,0 +1,25 @@
+import { get } from 'lodash';
+import type { AppState } from 'calypso/types';
+
+/**
+ * @param  {object}  state       Global state tree
+ * @param  {number}  productId   The productId to check for an intro offer
+ * @param  {number?}  siteId      The ID of the site we're querying
+ * @returns {boolean}        The raw price of intro offer, if available. null otherwise.
+ */
+export default function getIsEligibleForIntroOffer(
```
Comment (cbauerman) [inline client/state/selectors/get-is-eligible-for-intro-offer.ts]: This selector did not end up being used in this PR, but I left it in as there may be a need in the future.

File: client/state/sites/intro-offers/reducers.ts [hunk 3]
```diff
@@ -0,0 +1,61 @@
+import { Dictionary, keyBy } from 'lodash';
```
Comment (sirbrillig) [inline client/state/sites/intro-offers/reducers.ts]: Not a blocker, but I know the calypso team is trying to remove instances of lodash (there's even [a whole package](https://github.com/Automattic/wp-calypso/tree/trunk/packages/js-utils) devoted to it) and this PR adds a few. Is it possible to do these things without lodash and/or could we add functions to that package to avoid having to import lodash?

File: client/state/selectors/get-is-eligible-for-intro-offer.ts [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+	state: AppState,
+	productId: number,
+	siteId: number | 'none'
+): boolean {
+	const ineligibleReason = get(
```
Comment (sirbrillig) [inline client/state/selectors/get-is-eligible-for-intro-offer.ts]: This could probably be written without lodash, maybe like this?

```js
const ineligibleReason = state.sites.introOffers?.[ items ]?.[ siteId ]?.[ productId ]?.ineligibleReason ?? null;
```

File: client/state/selectors/get-intro-offers-request-status.ts [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+ *
+ * @param  {object}  state       Global state tree
+ * @param  {number}  siteId      The ID of the site we're querying
+ * @returns {string}             The request status or null if there has been no request
+ */
+export default function getIntroOfferRequestStatus(
+	state: AppState,
+	siteId?: number
+): RequestStatus | null {
+	return get( state.sites.introOffers, `${ siteId ?? 'none' }.requestStatus`, null );
```
Comment (sirbrillig) [inline client/state/selectors/get-intro-offers-request-status.ts]: I think you could write this without lodash like this:

```js
return state.sites.introOffers?.[ siteId ?? 'none' ]?.requestStatus ?? null;
```

File: client/state/selectors/get-is-requesting-into-offers.ts [hunk 6]
```diff
@@ -0,0 +1,14 @@
+import getIntroOfferRequestStatus from 'calypso/state/selectors/get-intro-offers-request-status';
+import { RequestStatus } from 'calypso/state/sites/intro-offers/types';
+import type { AppState } from 'calypso/types';
+/**
+ * Returns the status of the request and null if there has yet to be a request.
```
Comment (sirbrillig) [inline client/state/selectors/get-is-requesting-into-offers.ts]: Really minor since this is TS anyway, but the docs here are mostly wrong. The description says that it returns the status of the request or null, then strangely says "False otherwise", and it's return type is described as "string", but in fact this function returns a boolean.

File: client/state/selectors/get-is-eligible-for-intro-offer.ts [hunk 7]
```diff
@@ -0,0 +1,25 @@
+import { get } from 'lodash';
+import type { AppState } from 'calypso/types';
+
+/**
+ * @param  {object}  state       Global state tree
+ * @param  {number}  productId   The productId to check for an intro offer
+ * @param  {number?}  siteId      The ID of the site we're querying
+ * @returns {boolean}        The raw price of intro offer, if available. null otherwise.
```
Comment (sirbrillig) [inline client/state/selectors/get-is-eligible-for-intro-offer.ts]: Really minor but this function doesn't return `The raw price of intro offer, if available. null otherwise.`

Comment (cbauerman) [reply client/state/selectors/get-is-eligible-for-intro-offer.ts]: yeah, these are all copy/paste errors. We can get those cleaned up 👍.

File: client/state/sites/intro-offers/reducers.ts (hunk 3 above)
Comment (manzoorwanijk) [reply client/state/sites/intro-offers/reducers.ts]: Just curious about that decision...

Has the team considered a better alternative to lodash like [Ramda](https://ramdajs.com/)?
It's ultra lightweight and is designed for functional programming and makes creating and using predicates a breeze.

Comment (mattgawarecki) [reply client/state/sites/intro-offers/reducers.ts]: @manzoorwanijk You can find the original context for the initiative here: `p4TIVU-9Bf-p2` And there's always MGS if you want to dig deeper 😉

Comment (manzoorwanijk) [reply client/state/sites/intro-offers/reducers.ts]: Ah Thanks

File: client/components/data/query-intro-offers/index.tsx [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+const QueryIntroOffers: React.FC< OwnProps > = ( { siteId } ) => {
+	const dispatch = useDispatch();
+	const siteIdKey = siteId && typeof siteId === 'number' && siteId > 0 ? siteId : 'none';
+
+	const introOfferRequestStatus = useSelector( ( state ) =>
+		getIntroOfferRequestStatus( state, siteIdKey )
+	);
+
+	useEffect( () => {
+		if ( introOfferRequestStatus === null && siteIdKey !== undefined ) {
```
Comment (sirbrillig) [inline client/components/data/query-intro-offers/index.tsx]: > siteIdKey !== undefined

I'm pretty sure it's impossible for `siteIdKey` to be `undefined`, so we can remove this guard, unless you meant to check for `'none'`?

File: client/my-sites/checkout/composite-checkout/composite-checkout.tsx [hunk 9]
```diff
@@ -503,13 +508,15 @@ export default function CompositeCheckout( {
 		isInitialCartLoading ||
 		arePaymentMethodsLoading ||
 		paymentMethods.length < 1 ||
-		responseCart.products.length < 1;
+		responseCart.products.length < 1 ||
+		isLoadingIntroOffers;
```
Comment (sirbrillig) [inline client/my-sites/checkout/composite-checkout/composite-checkout.tsx:512]: This is probably fine due to the number of network requests happening during load, but in theory I think this could cause a flash of the cart before returning to the loading screen because `getIsIntroOfferRequesting` will return false if the request has not started yet, which will probably be the case when checkout first renders before the query component executes its fetch operation.

Fixing this in a safe way might be tricky, though, because while we could change `getIsIntroOfferRequesting` to return true if `getIntroOfferRequestStatus` is `null`, if there will never be a request (is that possible? Maybe if the cart is empty or there's no site selected? I'm not sure.) then we'd want this to be false.

Anyway, not a big deal but I thought I'd mention the possibility because I've seen this sort of confusion cause bugs in places before where some code was waiting for an API call to stop loading, thought it was done, but in fact the call hadn't actually started yet.

File: client/my-sites/checkout/composite-checkout/hooks/product-variants.tsx [hunk 10]
```diff
@@ -203,7 +203,7 @@ function isVariantAllowed(
 }
 
 function VariantPrice( { variant }: { variant: AvailableProductVariantAndCompared } ) {
-	const currentPrice = variant.priceFinal || variant.priceFull;
+	const currentPrice = variant.introductoryOfferPrice ?? variant.priceFinal ?? variant.priceFull;
```
Comment (sirbrillig) [inline client/my-sites/checkout/composite-checkout/hooks/product-variants.tsx]: I think I understand the intent here, but I'm a little confused about the data. This seems to be saying, "use the introductoryOfferPrice, but if that's not set, use the priceFinal, and if that's not set, use priceFull". However, it does this with nullish coalescing, and as far as the types say, `priceFinal` will never be `null` or `undefined`. So while the previous version of this line would display `priceFull` if `priceFinal` was `0`, the new version of this line is guaranteed to never display `priceFull` (unless the types are wrong). This seems like possibly a bug.

**Review by sirbrillig — COMMENTED** [review]
I tested this and it works as described! Nice work. I had a few comments inline, only one of which might be a blocker.

**Review by rcanepa — CHANGES_REQUESTED** [review]
I tested these changes thoroughly and found just one issue. On a site that has already used the introductory offer discount, the information on the Checkout page is not consistent, as shown in the following capture.

image

The item row says the discount is being applied, but the cart total says otherwise. Also, there is a message that says the order is not eligible for the introductory discount.

In short, the item row is not considering whether the discount has been already applied.

**Comment by cbauerman** [comment]
> I tested these changes thoroughly and found just one issue. On a site that has already used the introductory offer discount, the information on the Checkout page is not consistent, as shown in the following capture.
>
> [...]
> 
> The item row says the discount is being applied, but the cart total says otherwise. Also, there is a message that says the order is not eligible for the introductory discount.
> 
> In short, the item row is not considering whether the discount has been already applied.

@rcanepa this is consistent with how the cart works today. While this endpoint adds the information necessary to hide the notice, the focus of this PR is to make sure the discount is never an incorrect "Save 0%".

File: client/components/data/query-intro-offers/index.tsx (hunk 8 above)
Comment (cbauerman) [reply client/components/data/query-intro-offers/index.tsx]: Looks like a leftover from refactoring the Query component!

File: client/my-sites/checkout/composite-checkout/hooks/product-variants.tsx (hunk 10 above)
Comment (cbauerman) [reply client/my-sites/checkout/composite-checkout/hooks/product-variants.tsx]: > use the introductoryOfferPrice, but if that's not set, use the priceFinal, and if that's not set, use priceFull

Is definitely the intent. I think I got excited with the null coalescing operator. Since introductoryOfferPrice can be null ( in which case we use the original logic ) or 0 ( which is possible for free trial introductory offers ) we need something more like.

```js
const currentPrice =
		variant.introductoryOfferPrice !== null
			? variant.introductoryOfferPrice
			: variant.priceFinal || variant.priceFull;
```

**Comment by rcanepa** [comment]
> @rcanepa this is consistent with how the cart works today. While this endpoint adds the information necessary to hide the notice, the focus of this PR is to make sure the discount is never an incorrect "Save 0%".

Ok. It makes sense. Thanks for explaining that.
